cups/ipp-file.c

```diff
@@ -624,11 +624,71 @@ parse_value(_ipp_file_t      *f,	/* I  - IPP data file */
 	break;
 
     case IPP_TAG_STRING :
-        return (ippSetOctetString(ipp, attr, element, value, (int)strlen(value)));
+    	{
+    		if(value[0]=='<')				/* Input is binary(in form of hex) values*/
+    		{
+    			memmove(value, value+1, strlen(value));		/* Eliminate the '<' sign */
+    			char value_concat[50000];		/* Concatenated string with hexadecimal values without whitespace */
+    			int starting_line_number = f->linenum;
+    			while(value[strlen(value)-1] != '>') 
+    			{
+    				strcat(value_concat,value);
+    				if (!_ippFileReadToken(f, value, sizeof(value)))
+					  {
+					    report_error(f, v, user_data, "hexadecimal value not terminated starting line %d of \"%s\".", starting_line_number, f->filename);
+					    return (0);
+					  }
+    			}
+    			value[(strlen(value)-1)]='\0'; 		/* Eliminate the last '>' sign */
+    			strcat(value_concat,value);			/* Final concatenation for hexadecimal value to be complete*/
+    			int i;                         /* Iterating variable*/
+          for(i=0;i<strlen(value_concat);i++)		/* Sanity Check*/
+    			{
+    				if(!(value_concat[i] >= '0' && value_concat[i]<= '9') || (value_concat[i]>='a' && value_concat[i]<='f'))
+    				{
+    					report_error(f, v, user_data, "Bad hexadecimal value \"%s\" on line %d of \"%s\".", value_concat, starting_line_number, f->filename);
+	    				return (0);
+    				}
+    			}
+    			return (ippSetOctetString(ipp, attr, element, value_concat, (int)strlen(value_concat))); 
+    		}
+    		else
+    		{
+    			return (ippSetOctetString(ipp, attr, element, value, (int)strlen(value))); /* If a quoted string like "..", pass it on */
+    		}
+    		
+    	}
+        
         break;
 
     case IPP_TAG_TEXTLANG :
+    {
+    	(*attr)->values[element].string.text = _cupsStrAlloc(value);
+    	if (!_ippFileReadToken(f, value, sizeof(value)))	
+		  {
+		    report_error(f, v, user_data, "No Language Data in line %d of \"%s\".", f->linenum, f->filename);
+		    return (0);
+		  }
+		  memmove(value, value+1, strlen(value));
+		  value[strlen(value)-1]='\0';		/* Purge parenthesis */
+		(*attr)->values[element].string.language = _cupsStrAlloc(value);
+
+    }
+    break;
     case IPP_TAG_NAMELANG :
+    {
```

Comment:
Use the same code for IPP_TAG_NAMELANG and IPP_TAG_TEXTLANG (with the two case statements together feeding into the same code).